lib/aiken/rationals.ak

```diff
@@ -0,0 +1,628 @@
+//// This module implements operations between data structures of type Rational.
+////
+//// Rational numbers are represented as a pair of integers, a numerator and a denominator.
+////
+//// - the denominator is always positive
+//// - the sign is managed in the numerator.
+//// - the denominator is never 0.
+//// - the numerator and denominator are not explicitly reduced. This is done on demand.
+////
+//// Reducing should be done before write (e.g validating datum output)
+
+// todo: more comments (good docs)
+//      finish compare tests
+
+use aiken/math
+use aiken/option
+
+/// Useful for using Rationals in datums/redeemers which cannot have opaque types.
+/// 
+/// Use **check_rational** or **unsafe_check_rational** to convert to a Rational.
+pub type UncheckedRational {
+  numerator: Int,
+  denominator: Int,
+}
+
+/// Opaque type used to ensure the sign of the Rational is managed strictly in the numerator.
+pub opaque type Rational {
+  numerator: Int,
+  denominator: Int,
+}
+
+/// Converts an UncheckedRational to a Rational
+///
+/// ```aiken
+/// UncheckedRational { numerator: 2, denominator: 3 } |> check_rational() == Some(Rational { numerator: 2, denominator: 3 })
+/// ```
+pub fn check_rational(num: UncheckedRational) -> Option<Rational> {
+  let UncheckedRational { numerator, denominator } = num
+  ratio(numerator, denominator)
+}
+
+/// Unsafely converts an UncheckedRational to a Rational
+///
+/// ```aiken
+/// unsafe_check_rational(UncheckedRational { numerator: 2, denominator: 3 }) == Rational { numerator: 2, denominator: 3 }
+/// ```
+pub fn unsafe_check_rational(num: UncheckedRational) -> Rational {
+  let UncheckedRational { numerator, denominator } = num
+  unsafe_ratio(numerator, denominator)
+}
```

Comment:
```suggestion
```

Not in the stdlib please. Otherwise it completely defies the purpose of defining an abstraction.